docs(api): Revisions to the Examples page (#13322)
OpenTrons/opentrons-api: api/docs/v2/new_examples.rst

.. _new-examples:

*****************
Protocol Examples
*****************

> This page provides simple, ready-made protocols for Flex and OT-2. Feel free to copy and modify these examples to create unique protocols that help automate your laboratory workflows. Also, experimenting with these protocols is another way to build upon the skills you've learned from working through the :ref:`tutorial`. Try adding different hardware, labware, and commands to a sample protocol and test its validity after importing it into the Opentrons app.

Reviewer: ```suggestion
This page provides simple, ready-made protocols for Flex and OT-2. Feel free to copy and modify these examples to create unique protocols that help automate your laboratory workflows. Also, experimenting with these protocols is another way to build upon the skills you've learned from working through the :ref:`tutorial`. Try adding different hardware, labware, and commands to a sample protocol and test its validity after importing it into the Opentrons App.
```
🆙

Using These Protocols
=====================

> These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot. 

Reviewer: nitpick, this breaks really badly when formatted (all one line except `bot.`). maybe strike "liquid handling"

Author: @ecormany I don't see this break. There are 2 separate lines here, is that what you're referring to? I think this is ok.

Intro sentence:
> These sample protocols are designed for anyone using an Opentrons Flex or OT-2 liquid handling robot.

hard return, start next paragraph

> For our users with little to no Python experience, . . .

Reviewer: this, in the sandbox out-
put.

<img width="696" alt="image" src="https://github.com/Opentrons/opentrons/assets/1958332/4e46e0cd-dbf0-4ad0-9f2d-b0c0601b71c2">

Author: Fixing.

> For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use named arguments (or `named parameters <https://en.wikipedia.org/wiki/Named_parameter>`_) instead of positional arguments. For example::

Reviewer: I thought this was super-strict formatting, but you're right — Black will try to fit as much on one line even if that mixes and matches one-line vs. multi-line style.

But I think we should use Black parenthesis style, placing the closing parenthesis on a separate line, so:

```
    pipette_1 = protocol.load_instrument(
        instrument_name="flex_1channel_1000",
        mount="left",
        tip_racks=[tiprack_1]
    )
``` 

instead of 
```
        ...
        tip_racks=[tiprack_1])

Author: Oh, please, can we not do this? --- the dangling closing parens on it's own line.

We don't do this elsewhere in the docs. It looks weird and takes an extra line. I really think we shouldn't do this, unless it's part of a top to bottom, front to back revision of the code examples for the entire docs site.

Reviewer: ooooh, we have a mix elsewhere. yeah, we don't have to change it now. for a rainy day ☔

> For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use named arguments (or `named parameters <https://en.wikipedia.org/wiki/Named_parameter>`_) instead of positional arguments. For example::

Reviewer: ```suggestion
For our users with little to no Python experience, we’ve taken some liberties with the syntax and structure of the code to make it easier to understand. For example, we’ve formatted the samples with line breaks to show method arguments clearly and to avoid horizontal scrolling. Additionally, the methods use `named arguments <https://en.wikipedia.org/wiki/Named_parameter>`_ instead of positional arguments. For example::
```
i like the wiki link, and it includes "named arguments" in bold in the first line, so we don't have to equivocate here.

    # This code uses named arguments
    tiprack_1 = protocol.load_labware(
        load_name='opentrons_flex_96_tiprack_200ul',

>         location='D2')

Reviewer: here's an example of where i'd move the parenthesis down. i'm not going to mark all of them.

Author: I think we ought not to do this. The code works, we don't do it elsewhere, and it looks weird.

    # This code uses positional arguments
    tiprack_1 = protocol.load_labware('opentrons_flex_96_tiprack_200ul','D2')   

> Both examples instantiate the variable ``tiprack_1`` with a Flex tip rack, but the former is more explicit. It shows the parameter name and it's value together (e.g. ``location='D2'``), which may be helpful when you're unsure about what's going on in a protocol code sample.

Reviewer: ```suggestion
Both examples instantiate the variable ``tiprack_1`` with a Flex tip rack, but the former is more explicit. It shows the parameter name and its value together (e.g. ``location='D2'``), which may be helpful when you're unsure about what's going on in a protocol code sample.
```
😇

Python developers with more experience should feel free to ignore the code styling used here and work with these examples as you like.

Instruments and Labware
=======================

The sample protocols all use the following pipettes:

> * Flex 1-Channel Pipette (5-1000 µL). The API load name for this pipette is ``flex_1channel_1000``. 

Reviewer: ```suggestion
* Flex 1-Channel Pipette (5–1000 µL). The API load name for this pipette is ``flex_1channel_1000``. 
```
i like en dashes

> * P300 Single-Channel GEN2 pipette (20-300 µL) for the OT-2. The API load name for this pipette is ``p300_single_gen2``. 

Reviewer: ```suggestion
* P300 Single-Channel GEN2 pipette for the OT-2. The API load name for this pipette is ``p300_single_gen2``. 
```
the range isn't part of the product name here (it is for Flex pipettes)

                    location='D2')
                pipette_1 = protocol.load_instrument(
                    instrument_name='flex_1channel_1000',
                    mount='left',
                tip_racks=[tiprack_1])

                pipette_1.pick_up_tip()
                pipette_1.aspirate(100, plate['A1'])
                pipette_1.dispense(100, plate['B1'])
                pipette_1.drop_tip()

    .. tab:: OT-2

        .. code-block:: python
            :substitutions:

            from opentrons import protocol_api

            metadata = {'apiLevel': '2.15'}

            def run(protocol: protocol_api.ProtocolContext):
                plate = protocol.load_labware(
                load_name='corning_96_wellplate_360ul_flat',
                location=1)
            tiprack_1 = protocol.load_labware(
                load_name='opentrons_96_tiprack_300ul',
                location=2)
            p300 = protocol.load_instrument(
                instrument_name='p300_single',
                mount='left',
                tip_racks=[tiprack_1])

            p300.pick_up_tip()
            p300.aspirate(100, plate['A1'])
            p300.dispense(100, plate['B1'])

>             p300.drop_tip()

Reviewer: \+ 1 tab indent

Author: Yikes. Fixing.

Advanced Method
---------------

> This protocol accomplishes the same thing as the previous example, but does it a little more efficiently. Notice how it uses the :py:meth:`.InstrumentContext.transfer` method to move liquid between well plates. The ``aspirate`` and ``dispense`` arguments are part of ``transfer()`` method parameters. 

Reviewer: The arguments are `source` and `dest`, not aspirate and dispense.
Or, both aspirating and dispensing are performed as part of the transfer command.

Author: I think I mean that the actions of aspirating and dispensing were taken care of by `transfer` and you didn't need to call them.  My mistake. Fixing.

> The source and destination well arguments (e.g., `plate['A1'], plate['B1']`) are part of `transfer()` method parameters. You don’t need separate calls to `aspirate` or `dispense` here.

.. tabs::

    .. tab:: Flex

        .. code-block:: python
            :substitutions:

            from opentrons import protocol_api

            requirements = {'robotType': 'Flex', 'apiLevel': '|apiLevel|'}

            def run(protocol: protocol_api.ProtocolContext):
                plate = protocol.load_labware(
                    load_name='corning_96_wellplate_360ul_flat',
                    location='D1')
                tiprack_1 = protocol.load_labware(
                    load_name='opentrons_flex_96_tiprack_200ul',
                    location='D2')
                pipette_1 = protocol.load_instrument(
                    instrument_name='flex_1channel_1000',
                    mount='left',
                    tip_racks=[tiprack_1])
                # transfer 100 µL from well A1 to well B1
                pipette_1.transfer(100, plate['A1'], plate['B1'])

>                 pipette_1.drop_tip()

Reviewer: ```suggestion
```
Delete this line or protocol won't analyze. `transfer()` handles tip pickup and drop.

    
    .. tab:: OT-2

        .. code-block:: python
            :substitutions:

            from opentrons import protocol_api

            metadata = {'apiLevel': '2.15'}

            def run(protocol: protocol_api.ProtocolContext):
                plate = protocol.load_labware(
                    load_name='corning_96_wellplate_360ul_flat',
                    location=1)
            tiprack_1 = protocol.load_labware(
                    load_name='opentrons_96_tiprack_300ul',
                    location=2)
            p300 = protocol.load_instrument(
                instrument_name='p300_single',
                mount='left',
                tip_racks=[tiprack_1])
            # transfer 100 µL from well A1 to well B1

>             p300.transfer(volume=100, plate['A1'], plate['B1'])
>             p300.drop_tip()

Reviewer: ```suggestion
            p300.transfer(100, plate['A1'], plate['B1'])
```
Python error if positional argument follows named argument. Also removes drop tip for same reason as in Flex example.

            :substitutions:

            from opentrons import protocol_api

            metadata = {'apiLevel': '|apiLevel|'}

            def run(protocol: protocol_api.ProtocolContext):
                plate = protocol.load_labware(
                    load_name='corning_96_wellplate_360ul_flat',
                    location=1)
                tiprack_1 = protocol.load_labware(
                    load_name='opentrons_96_tiprack_300ul',
                    location=2)
                reservoir = protocol.load_labware(
                    load_name='usascientific_12_reservoir_22ml',
                    location=4)
                p300 = protocol.load_instrument(
                    instrument_name='p300_single',
                    mount='left',
                    tip_racks=[tiprack_1])
                
                # distribute 20 µL from reservoir:A1 -> plate:row:1
                # distribute 20 µL from reservoir:A2 -> plate:row:2
                # etc...
                # range() starts at 0 and stops before 8, creating a range of 0-7
                for i in range(8):
                    p300.distribute(200, reservoir.wells()[i], plate.rows()[i])

> Notice here how `Python's range class <https://docs.python.org/3/library/stdtypes.html#range>`_ (e.g., ``range(8)``) determines how many times the code loops. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well in the row, which is H1=7.

Reviewer: ```suggestion
Notice here how Python's :py:class:`range` class (e.g., ``range(8)``) determines how many times the code loops. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well in the row, which is H1=7.
```
Sphinx can directly reference the Python standard library.

    .. tab:: OT-2

        .. code-block:: python
            :substitutions:

            from opentrons import protocol_api

            metadata = {'apiLevel': '|apiLevel|'}

            def run(protocol: protocol_api.ProtocolContext):
                plate = protocol.load_labware(
                    load_name='corning_96_wellplate_360ul_flat',
                    location=1)
                tiprack_1 = protocol.load_labware(
                    load_name='opentrons_96_tiprack_300ul',
                    location=2)
                reservoir = protocol.load_labware(
                    load_name='usascientific_12_reservoir_22ml',
                    location=3)
                p300 = protocol.load_instrument(
                    instrument_name='p300_single', 
                    mount='right',
                    tip_racks=[tiprack_1])

                p300.pick_up_tip()

                # aspirate from the first 5 wells
                for well in reservoir.wells()[:4]:
                    p300.aspirate(volume=35, location=well)
                    p300.air_gap(10)
        
                p300.dispense(225, plate['A1'])

                p300.return_tip()

> Notice here how Python's `slice functionality <https://docs.python.org/3/library/functions.html#slice>`_ (in the code sample as ``[:4]``) lets us select the first five wells of the well plate only. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well used, which is E1=4. See also, the :ref:`tutorial-commands` section of the Tutorial.

Reviewer: ```suggestion
Notice here how Python's :py:class:`slice` functionality (in the code sample as ``[:4]``) lets us select the first five wells of the well plate only. Also, in Python, a range of numbers is *exclusive* of the end value and counting starts at 0, not 1. For the Corning 96-well plate used here, this means well A1=0, B1=1, C1=2, and so on to the last well used, which is E1=4. See also, the :ref:`tutorial-commands` section of the Tutorial.
```
sphinx-ify here too

> Dilution

Reviewer: we _could_ remove this section entirely because it's very much the basis for the tutorial. your call.

Author: I'd err on the side of keeping it.

========

> This protocol dispenses diluent to all wells of a Corning 96-well plate. Next, it dilutes eight samples from the reservoir across all 8 columns of the plate.

Reviewer: ```suggestion
This protocol dispenses diluent to all wells of a Corning 96-well plate. Next, it dilutes 8 samples from the reservoir across all 8 columns of the plate.
```
stick with numerals throughout since we have big and small numbers in one sentence

                    load_name='opentrons_96_tiprack_300ul',
                    location=2)
                tiprack_2 = protocol.load_labware(
                    load_name='opentrons_96_tiprack_300ul',
                    location=3)
            reservoir = protocol.load_labware(
                    load_name='usascientific_12_reservoir_22ml',
                    location=4)
            p300 = protocol.load_instrument(
                    instrument_name='p300_single',
                    mount='right',
                    tip_racks=[tiprack_1, tiprack_2])
            # Dispense diluent
            p300.distribute(50, reservoir['A12'], plate.wells())

            # loop through each row
            for i in range(8):
                # save the source well and destination column to variables
                source = reservoir.wells()[i]
                source = reservoir.wells()[i]
                row = plate.rows()[i]

            # transfer 30 µL of source to first well in column
            p300.transfer(30, source, row[0], mix_after=(3, 25))

            # dilute the sample down the column
            p300.transfer(
                30, row[:11], row[1:],

>                 mix_after=(3, 25))

Reviewer: +1 tab indent here

Author: oops. Fixing that now.

                    load_name='opentrons_flex_96_tiprack_200ul',
                    location='D3')
                reservoir = protocol.load_labware(
                    load_name='usascientific_12_reservoir_22ml',
                    location='C1')
                pipette_1 = protocol.load_instrument(
                    instrument_name='flex_1channel_1000',
                    mount='right',
                tip_racks=[tiprack_1, tiprack_2])

                # Volume amounts are for demonstration purposes only
                water_volumes = [
                    1,  2,  3,  4,  5,  6,  7,  8,
                    9,  10, 11, 12, 13, 14, 15, 16,
                    17, 18, 19, 20, 21, 22, 23, 24,
                    25, 26, 27, 28, 29, 30, 31, 32,
                    33, 34, 35, 36, 37, 38, 39, 40,
                    41, 42, 43, 44, 45, 46, 47, 48,
                    49, 50, 51, 52, 53, 54, 55, 56,
                    57, 58, 59, 60, 61, 62, 63, 64,
                    65, 66, 67, 68, 69, 70, 71, 72,
                    73, 74, 75, 76, 77, 78, 79, 80,
                    81, 82, 83, 84, 85, 86, 87, 88,
                    89, 90, 91, 92, 93, 94, 95, 96
                    ]

>                     pipette_1.distribute(water_volumes, reservoir['A12'], plate.wells())

Reviewer: ```suggestion
                pipette_1.distribute(water_volumes, reservoir['A12'], plate.wells())
```
-1 tab indent

                    load_name='opentrons_96_tiprack_300ul',
                    location=3)
                reservoir = protocol.load_labware(
                    load_name='usascientific_12_reservoir_22ml',
                    location=4)
                p300 = protocol.load_instrument(
                    instrument_name='p300_single', 
                    mount='right',
                    tip_racks=[tiprack_1, tiprack_2])

                # Volume amounts are for demonstration purposes only
                water_volumes = [
                    1,  2,  3,  4,  5,  6,  7,  8,
                    9,  10, 11, 12, 13, 14, 15, 16,
                    17, 18, 19, 20, 21, 22, 23, 24,
                    25, 26, 27, 28, 29, 30, 31, 32,
                    33, 34, 35, 36, 37, 38, 39, 40,
                    41, 42, 43, 44, 45, 46, 47, 48,
                    49, 50, 51, 52, 53, 54, 55, 56,
                    57, 58, 59, 60, 61, 62, 63, 64,
                    65, 66, 67, 68, 69, 70, 71, 72,
                    73, 74, 75, 76, 77, 78, 79, 80,
                    81, 82, 83, 84, 85, 86, 87, 88,
                    89, 90, 91, 92, 93, 94, 95, 96
                    ]

>                     p300.distribute(water_volumes, reservoir['A12'], plate.wells())

Reviewer: ```suggestion
                p300.distribute(water_volumes, reservoir['A12'], plate.wells())
```
outdent here too
